### DISCUSSION TITLE
[Variation font] How can I find the right face index or the right hb_variation_t for an name?

### DISCUSSION THREAD

**Discussion by moi15moi** [harfbuzz/harfbuzz#3882] [Q&A] [answered]
### What I want
I wanna map each font name from an variation font to an face_index.

For example, let's suppose I know my variation font contains these name:
- FamilyName
- FamilyName SemiBold
- FamilyName Bold

So, each name need to be "mapped" to an index.

### What's my problem

To do what, I need to get some information in the [Stat table](https://learn.microsoft.com/en-us/typography/opentype/spec/stat) and of course, the [fvar table](https://learn.microsoft.com/en-us/typography/opentype/spec/fvar).

I especially need these informations:
- [Axis Values](https://learn.microsoft.com/en-us/typography/opentype/spec/stat#axis-value-tables)
- [InstanceRecord](https://learn.microsoft.com/en-us/typography/opentype/spec/fvar#instancerecord)

I can get the **InstanceRecord** without any problem:
- [get_instance_count](https://github.com/harfbuzz/harfbuzz/blob/ae080bf202e08f3b9748a2d752ee5a1cc8a522f1/src/hb-ot-var-fvar-table.hh#L277)
- [get_instance](https://github.com/harfbuzz/harfbuzz/blob/ae080bf202e08f3b9748a2d752ee5a1cc8a522f1/src/hb-ot-var-fvar-table.hh#L408)

But, I can't get all the value I need for the **Axis Values**. 
I would need get the instance, because I also need the **AxisIndex** and the **Axis value**, because I can't map these to an InstanceRecord without those informations.
- [get_axis_value_count](https://github.com/harfbuzz/harfbuzz/blob/ae080bf202e08f3b9748a2d752ee5a1cc8a522f1/src/hb-ot-stat-table.hh#L511)
- [get_axis_value_name_id](https://github.com/harfbuzz/harfbuzz/blob/ae080bf202e08f3b9748a2d752ee5a1cc8a522f1/src/hb-ot-stat-table.hh#L513)

How could I get those information or is there already an function to do something similiar that I missed?

PS:
I know it it possible to support the [R/B/I/BI family model.](https://learn.microsoft.com/en-us/typography/opentype/spec/stat#alternate-font-family-models) with these methods but, this isn't what I want.:
- [hb_ot_name_get_utf16](https://harfbuzz.github.io/harfbuzz-hb-ot-var.html#hb-ot-var-named-instance-get-subfamily-name-id)
- [hb_ot_var_named_instance_get_subfamily_name_id](https://harfbuzz.github.io/harfbuzz-hb-ot-var.html#hb-ot-var-named-instance-get-subfamily-name-id)

**Comment by behdad** [comment]
> But, I can't get all the value I need for the **Axis Values**.

hb_ot_var_named_instance_get_design_coords

**Reply by moi15moi** [reply]
> > But, I can't get all the value I need for the **Axis Values**.
> 
> hb_ot_var_named_instance_get_design_coords

This isn't what I need.

I wanna have access to the [Axis Value table](https://learn.microsoft.com/en-us/typography/opentype/spec/stat#axis-value-tables).
I also need to be able to get the elidedFallbackNameID from the [Style Attributes Header](https://learn.microsoft.com/en-us/typography/opentype/spec/stat#style-attributes-header)

I need that, because each NamedInstance has their own name.
With DWrite, their name are not created by the subfamily_name_id [(hb_ot_var_named_instance_get_subfamily_name_id)](https://harfbuzz.github.io/harfbuzz-hb-ot-var.html#hb-ot-var-named-instance-get-subfamily-name-id)
It is created from the Axis value table.

Here is an example with Fonttools of what I am trying to do (Warning, only the naming with the Axis value table, format 1 has been implemented): https://gist.github.com/moi15moi/3f7704696aa16c64a2224888c032277b

**Comment by behdad** [comment, marked as answer]
Okay I see. You need the STAT table exposed in the API. Correct?

**Reply by moi15moi** [reply]
Exactly!

I think it is not currently available.

I think it could be a good feature to have one of these options:

1. Have access to the STAT table. It would make the process possible, but like you can see in my python script, it is complicated to find the right [Axis Values](https://learn.microsoft.com/en-us/typography/opentype/spec/stat#axis-value-tables) for an [InstanceRecord](https://learn.microsoft.com/en-us/typography/opentype/spec/fvar#instancerecord)
2. Add an method **hb_ot_var_named_instance_get_stat_name**. 
This method could return one of these:
a) The style string: "Bold condensed"
b) An array of each ordered nameID: [261, 262]

**Reply by moi15moi** [reply]
@behdad What do you think about my two previous suggestions?

**Reply by behdad** [reply]
> @behdad What do you think about my two previous suggestions?

I agree those are good ideas. Unfortunately I'm not very well-versed on the STAT table. I'll try to implement those.

**Reply by moi15moi** [reply]
Ok, thank you.
I am also not very well-versed on the STAT table. There is a lot of thing about the interaction between the different format of AxisValue that I don't know.

You may want to see what I have done here: https://gist.github.com/moi15moi/dd0fd510c03c7d80a274d69bf2edfb29
